docs/apply-live.md

```diff
@@ -0,0 +1,68 @@
+---
+nav_order: 6
+---
+
+# Architecture of apply-live
+{: .no_toc }
+
+1. TOC
+{:toc}
+
+## Copying into an "underlay"
+
+As noted in the architecture doc, everything in rpm-ostree is oriented
+around creating and managing hardlinked complete bootable filesystem trees.
+
+In this flow then, `rpm-ostree install --apply-live strace` will first
+create a new pending deployment, run sanity tests on it, prepare it to be booted, etc.
+
+However, the first time `apply-live` is invoked, we create an `overlayfs`
+mount over `/usr`.  It's mounted `ro` from the perspective of the rest
+of the system, but rpm-ostree can write to it.
+
+## Package and filesystem diffs
+
+When `apply-live` is invoked, rpm-ostree computes the diff between
+the source and target OSTree commit for `/usr`.  If this is the *first* `apply-live`,
+the source commit is the booted commit.  For subsequent invocations,
+it will be based on the current live commit.
+
+We also compute a package-level diff; this is how `apply-live`
+currently distinguishes between pure package additions versus upgrades.
+
+## Copying data for /usr
+
+Per the core OSTree model, almost everything we care about is in `/usr`.
+So the first step is to apply the diff to the transient writable `overlayfs`.
+
+One downside is that that this diff will take extra memory and disk space
+proportional to its size.
+
+## Updating /etc
+
+The second aspect we need to take care of is `/etc`.  Normally, the libostree
+core handles the `/etc` merge during shutdown as part of `ostree-finalize-staged.service`,
+but we need to do it now in order to ensure that we get new config files
+(or remove ones).
```

Comment:
Hmm, is there a gap here? If we fiddle with `/etc`, then the changes are not actually completely transient, and we may accumulate cruft in `/etc` which never goes away (e.g. if you then `rpm-ostree cleanup -p` because you actually just wanted to install a debug package you never intended to permanently overlay).

Reply:
Yep, this is absolutely an issue, but mostly mitigated by the fact that what we're stabilizing so far is `rpm-ostree install -A` in which case the admin is expressing the desire for the changes to persist.

If we added something like `rpm-ostree install --transient` this would be more of an issue.

I added a blurb on this.